wp/lib/bap_wp/src/compare.ml

```diff
@@ -56,21 +56,45 @@ let init_vars (env1 : Env.t) (env2 : Env.t) (vars : Var.Set.t)
   let inits2, env2 = Pre.init_vars vars env2 in
   (inits1 @ inits2), env1, env2
 
+(** [mk_smtlib2_compare] builds a constraint out of an smtlib2 string that can be used
+    as a comparison predicate between an original and modified binary. *)
+let mk_smtlib2_compare (env1 : Env.t) (env2 : Env.t) (smtlib_str : string) : Constr.t =
+  let var_map1 = (Env.get_var_map env1) in
+  let var_map2 = (Env.get_var_map env2) in
+  let smtlib_str = Env.EnvMap.fold var_map1 ~init:smtlib_str
+      ~f:(fun ~key:var ~data:z3_var smtlib_str ->
+          String.substr_replace_all smtlib_str ~pattern:((Var.name var) ^ "_orig") ~with_:(Z3.Expr.to_string z3_var)
+        ) in
+  let smtlib_str = Env.EnvMap.fold var_map2 ~init:smtlib_str
+      ~f:(fun ~key:var ~data:z3_var smtlib_str ->
+          String.substr_replace_all smtlib_str ~pattern:((Var.name var) ^ "_mod") ~with_:(Z3.Expr.to_string z3_var)
+        ) in
+  info "New smtlib string: %s \n" smtlib_str;
+  let declsym1 = Z3_utils.get_decls_and_symbols env1 in
+  let declsym2 = Z3_utils.get_decls_and_symbols env2 in
+  let declsym = declsym1 @ declsym2 in
+  let ctx = Env.get_context env1 in
+  Z3_utils.mk_smtlib2 ctx smtlib_str declsym
+
 let compare_blocks
     ~input:(input : Var.Set.t)
     ~output:(output : Var.Set.t)
     ~original:(blk1, env1 : Blk.t * Env.t)
     ~modified:(blk2, env2 : Blk.t * Env.t)
+    ~smtlib_post:(smtlib_post : string)
+    ~smtlib_hyp:(smtlib_hyp : string)
   : Constr.t * Env.t * Env.t =
   (* We only freshen variables in blk2, leaving those of blk1 with
      their original names. *)
   let env2 = Env.set_freshen env2 true in
   let output_eq_list, env1, env2 = set_to_eqs env1 env2 output in
-  let output_eq = Constr.mk_clause [] output_eq_list in
+  let smtlib_post = mk_smtlib2_compare env1 env2 smtlib_post in
+  let output_eq = Constr.mk_clause [] (smtlib_post :: output_eq_list) in
   let input_eq_list, env1, env2 = set_to_eqs env1 env2 input in
+  let smtlib_hyp = mk_smtlib2_compare env1 env2 smtlib_hyp in
   let pre1, _ = Pre.visit_block env1 output_eq blk1 in
   let pre2, _ = Pre.visit_block env2 pre1 blk2 in
-  let goal = Constr.mk_clause input_eq_list [pre2] in
+  let goal = Constr.mk_clause (smtlib_hyp :: input_eq_list) [pre2] in
```

Comment:
We probably don't want to add anything if the `smtlib_hyp` is trivial, since this may be a performance issue, if I remember correctly. Probably not that big a deal though.